File: app/shared/contracts/models/backup.py
"""Backup service contract models."""

from __future__ import annotations

from typing import Any, Literal

from pydantic import Field

from app.shared.contracts.registry import IOModel


class BackupModule:
    """Module identifier for Backup service."""

    NAME = "Backup"


class BackupMethods:
    """Full method identifiers for Backup service."""

    CREATE = f"{BackupModule.NAME}.Create"
    LIST = f"{BackupModule.NAME}.List"
    VERIFY = f"{BackupModule.NAME}.Verify"
    RESTORE = f"{BackupModule.NAME}.Restore"
    ROLLBACK = f"{BackupModule.NAME}.Rollback"
    HEALTH_CHECK = f"{BackupModule.NAME}.HealthCheck"


BackupComponentName = Literal["config", "db", "rag", "models"]
BackupComponentStatus = Literal["included", "skipped", "unavailable", "unsupported"]
BackupOperationStatus = Literal["ok", "denied", "failed", "unsupported", "not_found"]
BackupStorageKind = Literal["local", "s3", "gcs", "azure", "custom"]
BackupEncryptionMode = Literal["none", "passphrase", "age", "kms", "external"]


class BackupStorageTarget(IOModel):
    """Storage target for a backup operation.

    Secret material is intentionally represented as references or fingerprints;
    raw keys, passphrases, tokens, and connection strings must not be returned.
    """

    kind: BackupStorageKind = "local"
    uri: str | None = None
    encryption: BackupEncryptionMode = "none"
    key_ref: str | None = None
    credential_ref: str | None = None
    metadata: dict[str, Any] = Field(default_factory=dict)


class BackupComponentResult(IOModel):
    """One component's backup/restore status."""

    component: BackupComponentName
    status: BackupComponentStatus
    item_count: int | None = None
    bytes: int | None = None
    fingerprint: str | None = None
    redacted: bool = True
    message: str | None = None


class BackupServiceImpact(IOModel):
    """Service impact expected for restore or rollback."""

    service: str
    action: Literal["quiesce", "restart", "reload", "manual", "none"]
    required: bool = True
    reason: str


class BackupImpactPlan(IOModel):
    """Restore impact plan that clients must show before confirmation."""

    admin_critical: bool = True
    requires_quiesce: bool = True
    requires_restart: bool = True
    affected_services: list[BackupServiceImpact] = Field(default_factory=list)
    warnings: list[str] = Field(default_factory=list)


class BackupManifestSummary(IOModel):
    """UI-safe backup manifest summary."""

    backup_id: str
    created_at: str
    status: BackupOperationStatus
    storage: BackupStorageTarget
    components: list[BackupComponentResult] = Field(default_factory=list)
    manifest_digest: str
    schema_version: str = "aurora.backup.v1"
    encrypted: bool = False
    secrets_redacted: bool = True
    audit_receipt: str | None = None


class BackupCreateRequest(IOModel):
    """Create a backup manifest for selected backend state."""

    storage: BackupStorageTarget = Field(default_factory=BackupStorageTarget)
    components: list[BackupComponentName] = Field(
        default_factory=lambda: ["config", "db", "rag", "models"]
    )
    reason: str
    include_personal_data: bool = False
    correlation_id: str | None = None


class BackupCreateResponse(IOModel):
    """Created backup manifest response."""

    status: BackupOperationStatus
    backup: BackupManifestSummary | None = None
    audit_receipt: str
    message: str | None = None


class BackupListRequest(IOModel):
    """List known backup manifests."""

    limit: int = 50
    offset: int = 0
    include_failed: bool = True


class BackupListResponse(IOModel):
    """List of known backup manifests."""

    backups: list[BackupManifestSummary] = Field(default_factory=list)
    total: int = 0
    secrets_redacted: bool = True


class BackupVerifyRequest(IOModel):
    """Verify a backup manifest digest and component metadata."""

    backup_id: str
    storage: BackupStorageTarget | None = None


class BackupVerifyResponse(IOModel):
    """Backup verification result."""

    status: BackupOperationStatus
    backup_id: str
    verified: bool = False
    manifest_digest: str | None = None
    components: list[BackupComponentResult] = Field(default_factory=list)
    message: str | None = None


class BackupRestoreRequest(IOModel):
    """Restore or dry-run restore from a backup manifest."""

    backup_id: str
    storage: BackupStorageTarget | None = None
    components: list[BackupComponentName] | None = None
    dry_run: bool = True
    reason: str
    create_rollback: bool = True
    correlation_id: str | None = None


class BackupRestoreResponse(IOModel):
    """Restore result or dry-run impact plan."""

    status: BackupOperationStatus
    backup_id: str
    restored: bool = False
    rollback_backup_id: str | None = None
    impact_plan: BackupImpactPlan
    audit_receipt: str
    message: str | None = None


class BackupRollbackRequest(IOModel):
    """Rollback a prior restore using a rollback backup."""

    rollback_backup_id: str
    reason: str
    dry_run: bool = True
    correlation_id: str | None = None


class BackupRollbackResponse(IOModel):
    """Rollback result or dry-run impact plan."""

    status: BackupOperationStatus
    rollback_backup_id: str
    rolled_back: bool = False
    impact_plan: BackupImpactPlan
    audit_receipt: str
    message: str | None = None

File: packages/aurora-sdk/src/backup.ts
import type { AuroraClient } from './client.js'
import { routePath } from './descriptors.js'
import type { JsonObject } from './types.js'
import type { AuroraResponse } from './transport.js'

export const BACKUP_METHODS = {
  create: 'Backup.Create',
  list: 'Backup.List',
  verify: 'Backup.Verify',
  restore: 'Backup.Restore',
  rollback: 'Backup.Rollback',
  healthCheck: 'Backup.HealthCheck'
} as const

export type BackupComponentName = 'config' | 'db' | 'rag' | 'models'
export type BackupComponentStatus = 'included' | 'skipped' | 'unavailable' | 'unsupported'
export type BackupOperationStatus = 'ok' | 'denied' | 'failed' | 'unsupported' | 'not_found'
export type BackupStorageKind = 'local' | 's3' | 'gcs' | 'azure' | 'custom'
export type BackupEncryptionMode = 'none' | 'passphrase' | 'age' | 'kms' | 'external'

export interface BackupStorageTarget {
  kind: BackupStorageKind
  uri?: string | null
  encryption: BackupEncryptionMode
  key_ref?: string | null
  credential_ref?: string | null
  metadata: JsonObject
}

export interface BackupComponentResult {
  component: BackupComponentName
  status: BackupComponentStatus
  item_count?: number | null
  bytes?: number | null
  fingerprint?: string | null
  redacted: boolean
  message?: string | null
}

export interface BackupServiceImpact {
  service: string
  action: 'quiesce' | 'restart' | 'reload' | 'manual' | 'none'
  required: boolean
  reason: string
}

export interface BackupImpactPlan {
  admin_critical: boolean
  requires_quiesce: boolean
  requires_restart: boolean
  affected_services: BackupServiceImpact[]
  warnings: string[]
}

export interface BackupManifestSummary {
  backup_id: string
  created_at: string
  status: BackupOperationStatus
  storage: BackupStorageTarget
  components: BackupComponentResult[]
  manifest_digest: string
  schema_version: string
  encrypted: boolean
  secrets_redacted: boolean
  audit_receipt?: string | null
}

export interface BackupCreateRequest {
  storage?: BackupStorageTarget
  components?: BackupComponentName[]
  reason: string
  include_personal_data?: boolean
  correlation_id?: string | null
}

export interface BackupCreateResponse {
  status: BackupOperationStatus
  backup: BackupManifestSummary | null
  audit_receipt: string
  message?: string | null
}

export interface BackupListRequest {
  limit?: number
  offset?: number
  include_failed?: boolean
}

export interface BackupListResponse {
  backups: BackupManifestSummary[]
  total: number
  secrets_redacted: boolean
}

export interface BackupVerifyRequest {
  backup_id: string
  storage?: BackupStorageTarget | null
}

export interface BackupVerifyResponse {
  status: BackupOperationStatus
  backup_id: string
  verified: boolean
  manifest_digest?: string | null
  components: BackupComponentResult[]
  message?: string | null
}

export interface BackupRestoreRequest {
  backup_id: string
  storage?: BackupStorageTarget | null
  components?: BackupComponentName[] | null
  dry_run?: boolean
  reason: string
  create_rollback?: boolean
  correlation_id?: string | null
}

export interface BackupRestoreResponse {
  status: BackupOperationStatus
  backup_id: string
  restored: boolean
  rollback_backup_id?: string | null
  impact_plan: BackupImpactPlan
  audit_receipt: string
  message?: string | null
}

export interface BackupRollbackRequest {
  rollback_backup_id: string
  reason: string
  dry_run?: boolean
  correlation_id?: string | null
}

export interface BackupRollbackResponse {
  status: BackupOperationStatus
  rollback_backup_id: string
  rolled_back: boolean
  impact_plan: BackupImpactPlan
  audit_receipt: string
  message?: string | null
}

type BackupAdminPayload =
  | BackupCreateRequest
  | BackupVerifyRequest
  | BackupRestoreRequest
  | BackupRollbackRequest

export class BackupClient {
  constructor(private readonly client: AuroraClient) {}

  list(request: BackupListRequest = {}): Promise<AuroraResponse<BackupListResponse>> {
    return this.client.requestResult<BackupListResponse, BackupListRequest>(
      BACKUP_METHODS.list,
      request,
      { path: routePath('Backup', 'List') }
    )
  }

  create(request: BackupCreateRequest): Promise<AuroraResponse<BackupCreateResponse>> {
    return this.adminExecute<BackupCreateResponse>(BACKUP_METHODS.create, request, request.reason)
  }

  verify(request: BackupVerifyRequest, reason: string): Promise<AuroraResponse<BackupVerifyResponse>> {
    return this.adminExecute<BackupVerifyResponse>(BACKUP_METHODS.verify, request, reason)
  }

  restore(request: BackupRestoreRequest): Promise<AuroraResponse<BackupRestoreResponse>> {
    return this.adminExecute<BackupRestoreResponse>(BACKUP_METHODS.restore, request, request.reason)
  }

  rollback(request: BackupRollbackRequest): Promise<AuroraResponse<BackupRollbackResponse>> {
    return this.adminExecute<BackupRollbackResponse>(BACKUP_METHODS.rollback, request, request.reason)
  }

  private adminExecute<TData>(
    methodId: string,
    payload: BackupAdminPayload,
    reason: string
  ): Promise<AuroraResponse<TData>> {
    return this.client.result(async () => {
      const result = await this.client.admin.execute<TData>({
        methodId,
        payload: payload as unknown as JsonObject,
        reason,
        reauthConfirmed: true,
        affectedResources: ['admin.backups'],
        path: pathForBackupMethod(methodId)
      })
      return result.data
    })
  }
}

function pathForBackupMethod(methodId: string): string {
  const method = methodId.split('.')[1] ?? methodId
  return routePath('Backup', method)
}
